craft_parts/plugins/go_plugin.py

```diff
@@ -0,0 +1,151 @@
+        """Ensure the environment contains dependencies needed by the plugin.
+
+        :param part_dependencies: A list of the parts this part depends on.
+
+        :raises PluginEnvironmentValidationError: If the environment is invalid.
+        """
+        try:
+            version = self._execute("go version").strip()
+            if not version.startswith("go version"):
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason=f"invalid go compiler version {version!r}",
+                )
+            logger.debug("found %s", version)
+        except subprocess.CalledProcessError as err:
+            if err.returncode != PluginEnvironmentValidator.COMMAND_NOT_FOUND:
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason=f"go compiler failed with error code {err.returncode}",
+                ) from err
+
+            if part_dependencies is None:
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason="go compiler not found",
+                ) from err
+
+            if "go" not in part_dependencies:
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason=(
+                        f"go compiler not found and part {self._part_name!r} "
+                        f"does not depend on a part named 'go'"
+                    ),
+                ) from err
+
+
+class GoPlugin(Plugin):
+    """A plugin for go projects using go.mod.
+
+    The go plugin requires a go compiler installed on your system. This can
+    be achieved by adding the appropriate golang package to ``build-packages``,
+    or to have it installed or built in a different part. In this case, the
+    name of the part supplying the go compiler must be "go".
+
+    The go plugin uses the common plugin keywords as well as those for "sources".
+    Additionally, the following plugin-specific keywords can be used:
+
+    - ``go-buildtags``
+      (list of strings)
+      Tags to use during the go build. Default is not to use any build tags.
+    """
+
+    properties_class = GoPluginProperties
+    validator_class = GoPluginEnvironmentValidator
+
+    def get_build_snaps(self) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {"gcc"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {
+            "PARTS_GO_LDFLAGS": "-ldflags -linkmode=external",
+            "CGO_ENABLED": "1",
```

Comment:
We should probably remove these two. The first one is a Snapcraft workaround for classic confinement which is leaking into the pureness of this library :-)

The second one is more debatable and maybe the reason for gcc in the former.

Reply:
We can change the plugin to be pure go and address issues (possibly in a different way) as they appear.